### DISCUSSION TITLE
pipeline runs in a air-gapped kubernetes cluster  , workflow pod failed  because of  pulling image from internet registry

### DISCUSSION THREAD

**Discussion by lordebj** [kubeflow/pipelines#10677] [General]
i setup a air-gapped kubernetes cluster by using with kustomize build example |kubectl apply -f -
the environment is kubernetes 1.28.3 and kubeflow 1.8 
thing goes well,all the  pod run ok
but when i run the example pipeline 
the pod fail , it has to pull some image like 
 - image: gcr.io/ml-pipeline/kfp-driver@sha256:8e60086b04d92b657898a310ca9757631d58547e76bbbb8bfc376d654bef1707
  - image: gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance
  
 how can change it  to my private registry

**Comment by rimolive** [comment]
Launcher and driver images can be parametrized using KFP 2.0.5.

argoexec can be changed in the [workflow-controller command args](https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/third-party/argo/upstream/manifests/base/workflow-controller/workflow-controller-deployment.yaml#L34).

**Reply by lordebj** [reply]
thanks for you information. 

You mean when i setup the kubeflow by using kustomizie build  example|kubectl -f -  ,i  need to kustomize my workflow-controller  operator.
Can it be change after the  installing of kubeflow just by   kubectl delete -f  workflow-controller-deployment.yaml  and kubectl create -f  workflow-controller-deployment.yaml 

you information is about argoexec image, how about the kfp-driver  problem

another question:  the nobebook has the same problem,when i  new a notebook, the pod it needs  also switch to internet register.

**Comment by rimolive** [comment]
> you information is about argoexec image, how about the kfp-driver problem

This is a bit tricky because there's no mention in the manifests, but you can simply add the `V2_LAUNCHER_IMAGE` and the `V2_DRIVER_IMAGE` env vars in the `ml-pipeline` Deployment.

**Reply by lordebj** [reply]
i set the env like like 
kubectl set env deployment/ml-pipeline -n kubeflow V2_LAUNCHER_IMAGE="m.daocloud.io/gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance"

kubectl set env deployment/ml-pipeline -n kubeflow V2_DRIVER_IMAGE="m.daocloud.io/gcr.io/ml-pipeline/kfp-driver@sha256:8e60086b04d92b657898a310ca9757631d58547e76bbbb8bfc376d654bef1707"

 kubeclt get deploy ml-pipeline  -o yaml report  the env  is  ok
but when run the test pipe
it report:
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  51s   default-scheduler  Successfully assigned kubeflow-user-example-com/hello-pipeline-5vgsp-3986936489 to kf03.sinopec.com
  Normal  Pulling    50s   kubelet            Pulling image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance"
[root@kf01 ~]#

what is the problem.

**Reply by rimolive** [reply]
Looks like the pull is working, as I don't see any pull errors in the events.

**Comment by lordebj** [comment]
thanks.
the detail events :
Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  37m                    default-scheduler  Successfully assigned kubeflow-user-example-com/hello-pipeline-5vgsp-3986936489 to kf03.sinopec.com
  Warning  Failed     35m (x2 over 36m)      kubelet            Failed to pull image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": rpc error: code = Canceled desc = failed to pull and unpack image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": failed to resolve reference "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": failed to do request: Head "https://gcr.io/v2/ml-pipeline/argoexec/manifests/v3.3.10-license-compliance": context canceled
  Warning  Failed     33m (x4 over 36m)      kubelet            Error: ErrImagePull
  Warning  Failed     33m (x2 over 35m)      kubelet            Failed to pull image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": failed to pull and unpack image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": failed to resolve reference "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance": failed to do request: Head "https://gcr.io/v2/ml-pipeline/argoexec/manifests/v3.3.10-license-compliance": Forwarding failure
  Warning  Failed     33m (x6 over 36m)      kubelet            Error: ImagePullBackOff
  Normal   BackOff    7m46s (x116 over 36m)  kubelet            Back-off pulling image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance"
  Normal   Pulling    2m51s (x11 over 37m)   kubelet            Pulling image "gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance"


because i has not yes switched to my registry.  like   m.daocloud.io/gcr.io/ml-pipeline/argoexec:v3.3.10-license-compliance

**Reply by rimolive** [reply]
Let me know if you still need help. It looks like you found the issue and fixed?
